### PR TITLE
User ID support in .NET SDK

### PR DESCRIPTION
**Pull request by richa-d**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests if this modifies the Windows code?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+using Microsoft.AppCenter.Utils;
+using System;
+
+namespace Microsoft.AppCenter.Windows.Shared.Utils
+{
+    /**
+     * Utility to store and retrieve values for user identifiers.
+    */
+    public partial class UserIdContext
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]:
```suggestion
    public class UserIdContext
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        // The lock is static. Instance methods are not necessarily thread safe, but static methods are
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]:
```suggestion
```

That's not the reason we need a static lock. This is for the singleton instance.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        private static readonly object UserIdLock = new object();
+        private string _userId;
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: No need for a backing variable, we are doing nothing in get and set so no need for lock there.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Check if userId is valid for App Center.
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <returns>true if valid, false otherwise.</returns>
+        public static Boolean CheckUserIdValidForAppCenter(String userId)
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]:
```suggestion
        public static bool CheckUserIdValidForAppCenter(String userId)
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        /// <summary>
+        /// Current user identifier.
+        /// </summary>
+        public string UserId
+        {
+            get
+            {
+                lock (UserIdLock)
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: no need to lock, just declare UserId as

```
public string UserId { get; set; }
```

File: Tests/Microsoft.AppCenter.Test.Windows/Utils/UserIdContextTest.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        public void InitializeUserIdContextTest()
+        {
+           UserIdContext.Instance = new UserIdContext();
+        }
+
+        [TestMethod]
+        public void InstanceIsNotNull()
+        {
+            UserIdContext.Instance = null;
+            Assert.IsNotNull(UserIdContext.Instance);
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Test.Windows/Utils/UserIdContextTest.cs]: You can test singleton too.
Assign it to a local variable, then compare its Assert.Same as the second time you access it.

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Utils/ErrorLogHelper.cs [hunk 7]
```diff
@@ -175,7 +175,8 @@ private ManagedErrorLog InstanceCreateErrorLog(System.Exception exception)
                 Architecture = _processInformation.ProcessArchitecture,
                 Fatal = true,
                 Exception = CreateModelException(exception),
-                Sid = SessionContext.SessionId
+                Sid = SessionContext.SessionId,
+                UserId = UserIdContext.Instance.UserId
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Utils/ErrorLogHelper.cs]: Need to reflect in unit tests.

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Crashes.cs [hunk 8]
```diff
@@ -385,7 +385,7 @@ private void InstanceTrackError(System.Exception exception, IDictionary<string,
                     return;
                 }
                 properties = PropertyValidator.ValidateProperties(properties, "HandledError");
-                var log = new HandledErrorLog(exception: ErrorLogHelper.CreateModelException(exception), properties: properties, id: Guid.NewGuid(), device: null);
+                var log = new HandledErrorLog(exception: ErrorLogHelper.CreateModelException(exception), properties: properties, id: Guid.NewGuid(), device: null, userId: UserIdContext.Instance.UserId);
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Crashes.cs]: need to reflect in unit tests

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 1 above)
Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Done

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 2 above)
Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Done. Not using for UserId field.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 3 above)
Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Done

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 4 above)
Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Done

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 5 above)
Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Change to simple property

File: Tests/Microsoft.AppCenter.Test.Windows/Utils/UserIdContextTest.cs (hunk 6 above)
Comment (chandananjani2) [reply Tests/Microsoft.AppCenter.Test.Windows/Utils/UserIdContextTest.cs]: Added test

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Crashes.cs (hunk 8 above)
Comment (chandananjani2) [reply SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Crashes.cs]: Added unit test for two changes.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs (hunk 2 above)
Comment (guperrot) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: The comment is not relevant though, should just remove it as per suggestion above.

Comment (chandananjani2) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs]: Removed the comment

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Utils/ErrorLogHelper.cs (hunk 7 above)
Comment (richa-d) [reply SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Utils/ErrorLogHelper.cs]: Done.

File: Tests/Microsoft.AppCenter.Crashes.Test.Windows/HandledErrorTest.cs [hunk 9]
```diff
@@ -227,5 +227,32 @@ public void TrackErrorWhenDisabled()
...
+                {
+                    actualLog = JsonConvert.DeserializeObject<LogContainer>(content, LogSerializer.SerializationSettings).Logs.Single() as HandledErrorLog;
+                    semaphore.Release();
+                });
+            var exception = new System.Exception("Something went wrong.");
+            Crashes.TrackError(exception);
+
+            // Wait until the http layer sends the log.
+            semaphore.Wait(2000);
+            Assert.AreEqual(dummyUser, actualLog.UserId);
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Crashes.Test.Windows/HandledErrorTest.cs]: Add Assert.IsNotNull(actualLog) before that one.

File: Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs [hunk 10]
```diff
@@ -47,6 +47,9 @@ public void CreateErrorLog(string sessionId)
             // Set correlation identifier like Analytics does to set sessionId.
             SessionContext.SessionId = sessionId == null ? default(Guid?) : Guid.Parse(sessionId);
 
+            var mockUser = "dummyUser";
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs]: Let's use 2 more datarows and a new parameter userId

Then test

null, null
theExampleSid, null
null, theExampleUserId
both

File: Tests/Microsoft.AppCenter.Crashes.Test.Windows/HandledErrorTest.cs (hunk 9 above)
Comment (chandananjani2) [reply Tests/Microsoft.AppCenter.Crashes.Test.Windows/HandledErrorTest.cs]: Done

File: Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs (hunk 10 above)
Comment (chandananjani2) [reply Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs]: Done

File: Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs [hunk 11]
```diff
@@ -40,13 +40,16 @@ public void Cleanup()
         }
 
         [TestMethod]
-        [DataRow(null)]
-        [DataRow("da657080-2445-4b80-9b77-4798a6659954")]
-        public void CreateErrorLog(string sessionId)
+        [DataRow(null, null)]
+        [DataRow("da657080-2445-4b80-9b77-4798a6659954", null)]
+        [DataRow(null, "dummyUser")]
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Crashes.Test.Windows/Utils/ErrorLogHelperTest.cs]:
```suggestion
        [DataRow(null, "dummyUser")]
        [DataRow("da657080-2445-4b80-9b77-4798a6659954", "dummyUser")]
```

**Pull request merged**